src/coreclr/vm/loaderallocator.cpp

```diff
@@ -1876,6 +1885,70 @@ void AssemblyLoaderAllocator::CleanupHandles()
     }
 }
 
+void AssemblyLoaderAllocator::RegisterDependentHandleToNativeObjectForCleanup(LADependentHandleToNativeObject *dependentHandle)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_NOTRIGGER;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    _ASSERTE(dependentHandle != nullptr);
+
+    CrstHolder setLockHolder(&m_dependentHandleToNativeObjectSetCrst);
+
+    _ASSERTE(m_dependentHandleToNativeObjectSet.Lookup(dependentHandle) == NULL);
+    m_dependentHandleToNativeObjectSet.Add(dependentHandle);
+}
+
+void AssemblyLoaderAllocator::UnregisterDependentHandleToNativeObjectFromCleanup(LADependentHandleToNativeObject *dependentHandle)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    _ASSERTE(dependentHandle != nullptr);
+
+    CrstHolder setLockHolder(&m_dependentHandleToNativeObjectSetCrst);
+
+    _ASSERTE(m_dependentHandleToNativeObjectSet.Lookup(dependentHandle) != NULL);
+    m_dependentHandleToNativeObjectSet.Remove(dependentHandle);
+}
+
+void AssemblyLoaderAllocator::CleanupDependentHandlesToNativeObjects()
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_PREEMPTIVE;
+    }
+    CONTRACTL_END;
+
+    // Locks under which dependent handles may be used must all be taken here to ensure that a thread using a dependent handle
```

Comment:
This works though it seems a little awkward that each new CrossLAHash has to insert its lock here and all lock owners will need to work out their proper ordering relative to each other even if their locks would never otherwise be taken at the same time. Its not a big deal with only two users but if we start accruing more we might want to switch to a different pattern. One possibility would be adding a method to CrossLAHash traits which runs a lamda synchronized against all other accesses to that hash (presumably by doing it under the lock they would have added here). Then the DependentHandle could invoke that traits function passing a lamda that will delete the native object and clear the field. This lets each handle take the lock appropriate for that hash rather than taking all locks simultaneously.

In any case, I'm fine with this as-is, this was just thoughts for the future.

Reply:
Agreed it's not ideal. The lambda approach is interesting but it might have an issue with lock ordering. Currently `m_dependentHandleToNativeObjectSetCrst` is taken after the locks used for synchronizing access to the CrossLAHash instances. Since the handle cleanup code has to iterate the set of handles, it would have to first take `m_dependentHandleToNativeObjectSetCrst` and then the traits method would acquire its lock in the opposite order. Still thinking about ways to improve it, but decided to keep it simple for now.

Reply:
Yeah I noticed that and handwaved a bit that we could make a copy of the list before iterating it, or perhaps enter the lock to set a flag indicating that the LoaderAllocator is cleaning up so no new handles could be added. Happy to leave it how it is for now.

Reply:
I agree its awkward, but I do not expect there to be many of these hashtables. They are quite expensive and complicated. I do not believe its worthwhile to invest in improving an api that works well enough, as it is.